### PR TITLE
Implementing an URL object

### PR DESCRIPTION
**Pull request by burzum**

### CODE REVIEW

File: src/Routing/Url.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Routing;
+
+/**
+ * URL Object
```
Comment (markstory) [inline src/Routing/Url.php]: Some context on how/when to use this object would be good given that this could be something developers interact with very often.

File: src/Routing/Url.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Sets a query param
+     *
+     * @param string $key Query param name
+     * @param string|int|float $value Value
+     * @return $this
+     */
+    public function setQuery($key, $value)
+    {
```
Comment (markstory) [inline src/Routing/Url.php]: How do custom routing parameters get set?

File: src/Routing/Url.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Routing;
+
+/**
+ * URL Object
+ */
+class Url
```
Comment (robertpustulka) [inline src/Routing/Url.php]: Maybe this should implement `ArrayAccess`?

File: src/Routing/Url.php (hunk 2 above)
Comment (ADmad) [inline src/Routing/Url.php]: I guess a new method `setParam($key, $value)` would need to be added.

**Comment by ADmad** [comment]
Do we need `with` prefixed immutable version of all these methods? :trollface:

File: src/Routing/Url.php [hunk 4]
```diff
@@ -175,4 +181,59 @@ public function __toString()
...
+    /**
+     * Offset to unset
+     *
+     * @link http://php.net/manual/en/arrayaccess.offsetunset.php
+     * @param mixed $offset The offset to unset.
+     * @return void
+     */
+    public function offsetUnset($offset)
+    {
+        unset($offset);
```
Comment (robertpustulka) [inline src/Routing/Url.php]: Should be  `unset($this->url[$offset]);`

File: src/Routing/Url.php (hunk 2 above)
Comment (burzum) [inline src/Routing/Url.php]: There is already setPass()? I could rename it?

Comment (markstory) [inline src/Routing/Url.php]: But setPass() only sets 'passed' arguments. I'm thinking about custom routing params like `/blog/:title`. How would someone set `title`?

Comment (burzum) [inline src/Routing/Url.php]: Ahh... I totally forgot them... 🙄

Comment (burzum) [inline src/Routing/Url.php]: Done.

**Comment by burzum** [comment]
@markstory how about setting plugin and prefix by default to null and false? This has been a common annoying thing in the past when declaring URLs in navigation items. When working in a plugin environment and you haven't set it, you automatically ended up with a link to the plugin.

**Comment by dereuromark** [comment]
This change in behavior could be quite the heavy BC break for some.. While I also found it annoying I learned to live with it. And use my own convenience methods here instead.
If changed, this could break quite a few existing plugins relying on the current behavior however.

**Comment by burzum** [comment]
@dereuromark how would it break if you don't explicitly use the URL object? :)

File: src/Routing/Url.php [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        return $this;
+    }
+
+    /**
+     * Sets values to be passed in the URL
+     *
+     * @param string $pass Args to be passed to a controller action
+     * @return $this
+     */
+    public function setPass($pass)
```
Comment (robertpustulka) [inline src/Routing/Url.php]: Should this be named `addPass`? `setPass` suggests that the old value would be overwritten.

File: src/Routing/Url.php [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        return $this;
+    }
+
+    /**
+     * Absolute URL
+     *
+     * @param bool $absolute Generate an absolute URL or not
+     * @return $this
+     */
+    public function absolute($absolute = true)
```
Comment (robertpustulka) [inline src/Routing/Url.php]: `setAbsolute` ?

Comment (dereuromark) [inline src/Routing/Url.php]: Agreed, because we also need getters here usually.

File: src/Routing/Url.php [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        return $this;
+    }
+
+    /**
+     * Sets values to be passed in the URL
+     *
+     * @param string $pass Args to be passed to a controller action
+     * @return $this
+     */
+    public function addPass($pass)
```
Comment (ravage84) [inline src/Routing/Url.php]: ``addPassedValues()``?

File: src/Routing/Url.php [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        foreach ($params as $key => $value) {
+            $this->setParam($key, $value);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Sets a query param
```
Comment (ravage84) [inline src/Routing/Url.php]: param => parameter. Can we generally write out such wods? I guess that wouldn't hurt.

File: src/Routing/Url.php [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @param string $key Query param name
+     * @param string|int|float $value Value
```
Comment (ravage84) [inline src/Routing/Url.php]: Query parameter value.

File: src/Routing/Url.php [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+            $this->url['?'] = [];
+        }
+
+        $this->url['?'] = array_merge($this->url['?'], $params);
+
+        return $this;
+    }
+
+    /**
+     * Absolute URL
```
Comment (ravage84) [inline src/Routing/Url.php]: [Set w|W]hether to generate an absolute URL or not.

File: src/Routing/Url.php [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+     */
+    public function setAbsolute($absolute = true)
+    {
+        $this->absolute = (bool)$absolute;
+
+        return $this;
+    }
+
+    /**
+     * To string
```
Comment (ravage84) [inline src/Routing/Url.php]: Get URL as string.

File: src/Routing/Url.php [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @return string;
+     */
+    public function toString()
+    {
+        return Router::url($this->url, $this->absolute);
+    }
+
+    /**
+     * To array
```
Comment (ravage84) [inline src/Routing/Url.php]: Get URL as array.

File: src/Routing/Url.php [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        return $this->url;
+    }
+
+    /**
+     * To string
```
Comment (ravage84) [inline src/Routing/Url.php]: "To String auto convertion" or "To String Magic Method"?

File: src/Routing/Url.php [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @return string String URL
```
Comment (ravage84) [inline src/Routing/Url.php]: URL as string.

File: src/Routing/Url.php [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+     */
+    public function toString()
+    {
+        return Router::url($this->url, $this->absolute);
+    }
+
+    /**
+     * To array
+     *
+     * @return array
```
Comment (ravage84) [inline src/Routing/Url.php]: URL as array.

File: src/Routing/Url.php [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        $this->absolute = (bool)$absolute;
+
+        return $this;
+    }
+
+    /**
+     * To string
+     *
+     * @return string;
```
Comment (ravage84) [inline src/Routing/Url.php]: URL as string.

File: src/Routing/Url.php [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+     */
+    public function __toString()
+    {
+        return $this->toString();
+    }
+
+    /**
+     * Whether a offset exists
+     *
+     * @link http://php.net/manual/en/arrayaccess.offsetexists.php
```
Comment (ravage84) [inline src/Routing/Url.php]: May be after the ``@return``?

File: src/Routing/Url.php [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+            return $this->url[$offset];
+        }
+
+        return null;
+    }
+
+    /**
+     * Offset to set
+     *
+     * @link http://php.net/manual/en/arrayaccess.offsetset.php
```
Comment (ravage84) [inline src/Routing/Url.php]: May be after the ``@return``?

File: src/Routing/Url.php [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+     */
+    public function offsetSet($offset, $value)
+    {
+        $this->url[$offset] = $value;
+    }
+
+    /**
+     * Offset to unset
+     *
+     * @link http://php.net/manual/en/arrayaccess.offsetunset.php
```
Comment (ravage84) [inline src/Routing/Url.php]: May be after the ``@return``?

**Review by ravage84 — COMMENTED** [review]
Nice! Thanks for taking care of it!

File: src/Routing/Url.php (hunk 18 above)
Comment (inoas) [inline src/Routing/Url.php]: What should offset be for?

File: src/Console/CommandRunner.php [hunk 20]
```diff
@@ -20,6 +20,7 @@
 use Cake\Console\Exception\StopException;
 use Cake\Console\Shell;
 use Cake\Core\ConsoleApplicationInterface;
+use Cake\Core\HttpApplicationInterface;
```
Comment (burzum) [inline src/Console/CommandRunner.php]: No but I'll take care of it as well.

File: src/Routing/Url.php (hunk 19 above)
Comment (burzum) [inline src/Routing/Url.php]: Do we have a convention for that? I usually put things like that before params and returns.

File: src/Routing/Url.php [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        $this->url['action'] = $action;
+
+        return $this;
+    }
+
+    /**
+     * Sets the controller
+     *
+     * @param string $controller Controller class name without the controller suffix
```
Comment (burzum) [inline src/Routing/Url.php]: What if you have a case like `AccountingControllerController`? ;)

**Comment by inoas** [comment]
What about separating cake objects that store state into two categories:
1.) read/write objects, like say: request/response or entities
2.) builders, objects mostly used to set multiple values and then be consumed, like say: query builder or routing/url options.

Then for 1. we can use `get*()` and `set*()` and where for 2. we can use `get*()` for getting (if at all feasible) and the setters would just be the method names.

This would reduce verbosity but still allow IDE completion and strict method signatures and return types, wouldn't it? What would the drawbacks be?

**Comment by burzum** [comment]
@markstory is there something left before this can be merged?

File: src/Routing/Url.php [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    /**
+     * Sets a query parameter
+     *
+     * @param string $key Query Parameter name
+     * @param string|int|float $value Parameter value
+     * @return $this
+     */
+    public function setQuery($key, $value)
```
Comment (ADmad) [inline src/Routing/Url.php]: This method doesn't seem to have test coverage.

File: src/Routing/Url.php (hunk 7 above)
Comment (ADmad) [inline src/Routing/Url.php]: This method doesn't seem to have test coverage.

**Comment by ADmad** [comment]
I kinda like @ionas' suggestion. For builders which are primarily going to be used to set vars (this one doesn't even have any getter methods) adding "set" prefix to the setters adds unnecessary verbosity.

**Comment by robertpustulka** [comment]
@burzum isn't the merge with 3.next messed up here?

File: src/Console/CommandRunner.php [hunk 23]
```diff
@@ -14,11 +14,7 @@
  */
 namespace Cake\Console;
 
-use Cake\Console\CommandCollection;
-use Cake\Console\CommandCollectionAwareInterface;
-use Cake\Console\ConsoleIo;
 use Cake\Console\Exception\StopException;
-use Cake\Console\Shell;
```
Comment (robertpustulka) [inline src/Console/CommandRunner.php]: Why are those removed?

**Comment by chinpei215** [comment]
Replaced by #11036.
